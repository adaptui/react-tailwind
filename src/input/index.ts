export * from "./__utils";
export * from "./Input";
export * from "./InputBase";
export * from "./InputPrefix";
export * from "./InputProps";
export * from "./InputSuffix";
export * from "./InputUIState";
export * from "./InputWrapper";
