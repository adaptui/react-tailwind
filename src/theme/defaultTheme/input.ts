export const input = {
  wrapper: "relative inline-flex shadow-sm",
  base:
    "appearance-none w-full outline-none px-2 py-1 text-xs font-medium text-gray-500 border h-6 border-gray-300 hover:border-gray-400 rounded-md focus:text-gray-800 focus:border-blue-500 focus:shadow-input",
  disabled: "bg-gray-100",
  invalid: "border-red-300 hover:border-red-400",
  children: {
    base:
      "flex items-center justify-center text-xs absolute inset-y-0 h-6 px-2 text-gray-800 bg-transparent border border-transparent rounded-md focus:outline-none focus:border-blue-500 focus:shadow-input ",
    prefix: "left-0",
    suffix: "right-0",
  },

  group: {
    base: "inline-flex items-center justify-center rounded-md",
    children: {
      base:
        "inline-flex items-center bg-gray-100 h-6 text-xs font-medium text-gray-800 px-2 border border-gray-300 shadow-sm focus:outline-none focus:border-blue-500 focus:shadow-input",
      prefix: "-mr-px rounded-l-md",
      suffix: "-ml-px rounded-r-md",
    },
  },
};
