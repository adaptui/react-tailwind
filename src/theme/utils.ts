import { defaults, isString, isUndefined, mergeWith, tcm } from "../utils";

import {
  DefaultTheme,
  ExtendableDefaultTheme,
  PartialDefaultTheme,
} from "./types";

export const extendTheme = <T = ExtendableDefaultTheme>(theme: T): T => {
  return theme;
};

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
function collectExtends(items: ExtendableDefaultTheme[]) {
  return items.reduce((merged, { extend }) => {
    return mergeWith(merged, extend, (mergedValue, extendValue) => {
      if (isUndefined(mergedValue)) {
        return [extendValue];
      }

      if (Array.isArray(mergedValue)) {
        return [extendValue, ...mergedValue];
      }

      return [extendValue, mergedValue];
    });
  }, {});
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeThemes(themes: PartialDefaultTheme[]) {
  return {
    ...themes.reduce<DefaultTheme>(
      (merged, theme) => defaults(merged, theme),
      {} as DefaultTheme,
    ),

    // In order to resolve n config objects, we combine all of their `extend` properties
    // into arrays instead of objects so they aren't overridden.
    extend: collectExtends(themes),
  };
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeExtensions({
  extend,
  ...theme
}: DefaultTheme & { extend?: PartialDefaultTheme }) {
  return mergeWith(theme, extend, (themeValue, extendValue) => {
    return mergeWith(
      themeValue,
      ...extendValue,
      (merged: any, value: any[]) => {
        if (isString(merged) && isString(value)) {
          return tcm(merged, value);
        }

        return undefined;
      },
    );
  });
}
