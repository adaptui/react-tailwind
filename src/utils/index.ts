export * from "./assertions";
export * from "./createContext";
export * from "./debounce";
export * from "./global-types";
export * from "./object";
export * from "./react";
export * from "./tailwindMerge";
export * from "./types";
