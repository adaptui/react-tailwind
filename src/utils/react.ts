import * as React from "react";

import { isFunction } from "./assertions";
import { cx } from "./tailwindMerge";
import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};
