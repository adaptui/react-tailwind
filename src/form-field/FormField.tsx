import React from "react";
import { cx } from "@renderlesskit/react";

import { useTheme } from "../theme";
import { useId } from "../hooks/useId";
import { Box, BoxProps } from "../box";
import { useBoolean } from "../hooks/useBoolean";
import { useFormControl } from "./useFormControl";
import { createContext, runIfFn } from "../utils";
import { forwardRefWithAs, RenderProp } from "../utils/types";

export type CommonFieldProps = {
  id?: string;
  isDisabled?: boolean;
  isInvalid?: boolean;
  isReadOnly?: boolean;
  isRequired?: boolean;
};

type FormFieldContextProps = Omit<
  ReturnType<typeof useFormContextValues>,
  "htmlProps"
>;

const [
  FormFieldContextProvider,
  useFormFieldContext,
] = createContext<FormFieldContextProps>({ strict: false });

export { useFormFieldContext };

const useFormContextValues = (props: FormFieldProps) => {
  const {
    id,
    isRequired,
    isInvalid,
    isDisabled,
    isReadOnly,
    ...htmlProps
  } = props;
  const uuid = useId(id);
  const idProp = `field-${uuid}`;

  const labelId = `${idProp}-label`;
  const errorTextId = `${idProp}-error-text`;
  const helpTextId = `${idProp}-help-text`;

  const [hasHelpText, setHasHelpText] = useBoolean(false);
  const [hasErrorText, setHasErrorText] = useBoolean(false);

  const context = {
    isRequired,
    isInvalid,
    isReadOnly,
    isDisabled,
    inputId: id,
    fieldId: idProp,
    labelId,
    errorTextId,
    helpTextId,
    hasHelpText,
    hasErrorText,
    setHasHelpText,
    setHasErrorText,
    htmlProps,
  };

  return context;
};

export type FormFieldProps = BoxProps & CommonFieldProps;

export const FormField = forwardRefWithAs<
  FormFieldProps &
    RenderProp<
      FormFieldContextProps & {
        inputProps: ReturnType<typeof useFormControl>;
      }
    >,
  HTMLDivElement,
  "div"
>((props, ref) => {
  const { children, className, ...rest } = props;

  const theme = useTheme();
  const { htmlProps, ...context } = useFormContextValues(rest);

  const formFieldStyles = cx(theme.formField.wrapper.base, className);

  return (
    <FormFieldContextProvider value={context}>
      <Box
        ref={ref}
        role="group"
        id={context.fieldId}
        className={formFieldStyles}
        {...htmlProps}
      >
        <FormRenderProp children={children} htmlProps={htmlProps} />
      </Box>
    </FormFieldContextProvider>
  );
});

const FormRenderProp: React.FC<{ htmlProps: FormFieldProps }> = ({
  children,
  htmlProps,
}) => {
  const context = useFormFieldContext();
  const controlProps = useFormControl(htmlProps);

  return (
    <>
      {runIfFn(children, {
        ...context,
        inputProps: controlProps,
      })}
    </>
  );
};
