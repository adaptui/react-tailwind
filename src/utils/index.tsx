import * as React from "react";
import { createTailwindMerge } from "tailwind-merge";

import { cx } from "../utils";

import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

export interface CreateContextOptions {
  /**
   * If `true`, React will throw if context is `null` or `undefined`
   * In some cases, you might want to support nested context, so you can set it to `false`
   */
  strict?: boolean;
  /**
   * Error message to throw if the context is `undefined`
   */
  errorMessage?: string;
  /**
   * The display name of the context
   */
  name?: string;
}

export type CreateContextReturn<T> = [
  React.Provider<T>,
  () => T,
  React.Context<T>,
];

/**
 * Creates a named context, provider, and hook.
 *
 * @param options create context options
 */
export function createContext<ContextType>(options: CreateContextOptions = {}) {
  const {
    strict = true,
    errorMessage = "useContext: `context` is undefined. Seems you forgot to wrap component within the Provider",
    name,
  } = options;

  const Context = React.createContext<ContextType | undefined>(undefined);

  Context.displayName = name;

  function useContext() {
    const context = React.useContext(Context);

    if (!context && strict) {
      throw new Error(errorMessage);
    }

    return context;
  }

  return [
    Context.Provider,
    useContext,
    Context,
  ] as CreateContextReturn<ContextType>;
}

// Function assertions
export function isFunction(value: any): value is Function {
  return typeof value === "function";
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// SSR check
export function canUseDOM() {
  return !!(
    typeof window !== "undefined" &&
    window.document &&
    window.document.createElement
  );
}

export const isBrowser = canUseDOM();

// Array assertions
export function isArray<T>(value: any): value is Array<T> {
  return Array.isArray(value);
}

// Object assertions
export const isObject = (value: any): value is Dict => {
  const type = typeof value;

  return value != null && type === "object" && !isArray(value);
};

export const objectKeys = <T extends Dict>(obj: T) =>
  Object.keys(obj) as unknown as (keyof T)[];

// Corresponds to 10 frames at 60 Hz.
// A few bytes payload overhead when lodash/debounce is ~3 kB and debounce ~300 B.
export const debounce = (func: Function, wait = 166) => {
  let timeout: any;
  function debounced(this: any, ...args: any) {
    // eslint-disable-next-line consistent-this
    const that = this;
    const later = () => {
      func.apply(that, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  }

  debounced.clear = () => {
    clearTimeout(timeout);
  };

  return debounced;
};

// String assertions
export function isString(value: any): value is string {
  return Object.prototype.toString.call(value) === "[object String]";
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export function isUndefined(value: any): value is undefined {
  return typeof value === "undefined" || value === undefined;
}

export const tcm = createTailwindMerge(getDefaultConfig => {
  const defaultConfig = getDefaultConfig();

  return {
    ...defaultConfig,
    classGroups: {
      ...defaultConfig.classGroups,
      "font-size": [
        ...defaultConfig.classGroups["font-size"],
        // Defined custom text presets
        {
          text: ["cxs", "paragraph-cxs", "paragraph-sm", "2base"],
        },
      ],
    },
  };
});

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

export * from "./types";
export { cx } from "@renderlesskit/react";
