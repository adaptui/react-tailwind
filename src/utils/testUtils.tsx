import {
  render as RtlRender,
  RenderOptions,
  waitFor,
  RenderResult,
} from "@testing-library/react";
import * as React from "react";
import { RunOptions } from "axe-core";
import { configureAxe } from "jest-axe";

// @ts-ignore
import tailwindConfig from "../../tailwind.config";
import { RenderlesskitProvider } from "../theme";
export * from "@testing-library/react";

type Render = (
  children: React.ReactNode,
  options?: RenderOptions,
) => RenderResult;

export const render: Render = (children, options = {}) => {
  return RtlRender(
    <RenderlesskitProvider tailwindConfig={tailwindConfig}>
      {children}
    </RenderlesskitProvider>,
    options,
  );
};

export const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

type UI = Parameters<typeof render>[0] | Element;
type TestA11YOptions = RenderOptions & {
  axeOptions?: RunOptions;
};

export const testA11y = async (
  ui: UI,
  { axeOptions, ...options }: TestA11YOptions = {},
) => {
  jest.useRealTimers();

  const container = React.isValidElement(ui)
    ? render(ui, options).container
    : ui;

  // @ts-ignore
  const results = await axe(container, axeOptions);

  expect(results).toHaveNoViolations();
};

const DELAY = 0;
const LOAD_IMAGE = "load.png";
const ERROR_IMAGE = "error.png";
const orignalImage = window.Image;

export const mockImage = (loadState: string) => {
  jest.useFakeTimers();

  (window.Image as unknown) = class MockImage {
    onload: () => void = () => {};
    onerror: () => void = () => {};
    src: string = "";
    constructor() {
      if (loadState === LOAD_IMAGE) {
        setTimeout(() => {
          this.onload();
        }, DELAY);
      }
      if (loadState === ERROR_IMAGE) {
        setTimeout(() => {
          this.onerror();
        }, DELAY);
      }
      return this;
    }
  };
};

mockImage.DELAY = DELAY;
mockImage.LOAD_IMAGE = LOAD_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.load = () => mockImage(LOAD_IMAGE);
mockImage.error = () => mockImage(ERROR_IMAGE);
mockImage.restoreMock = () => (window.Image = orignalImage);
mockImage.advanceTimer = async () => {
  return await waitFor(() => {
    jest.advanceTimersByTime(mockImage.DELAY);
  });
};
