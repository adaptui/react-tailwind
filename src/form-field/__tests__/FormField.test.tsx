import * as React from "react";
import { screen } from "@testing-library/react";

import {
  FormField,
  FormLabel,
  FormErrorText,
  FormHelperText,
  FormRequiredText,
} from "../index";
import { Input } from "../../input";
import { FormFieldProps } from "../FormField";
import { render, testA11y } from "../../utils/testUtils";

const LABEL_TEXT = "Form label";
const HINT_TEXT = "This is hint text";
const ERROR_TEXT = "This is error text";
const REQUIRED_TEXT = "Required";
const LABEL_TESTID = "testid-label";
const HINT_TESTID = "testid-hint";
const ERROR_TESTID = "testid-error";
const REQUIRED_TESTID = "testid-required";
const INPUT_TESTID = "testid-input";

const FormFieldComponent: React.FC<FormFieldProps> = props => {
  return (
    <FormField {...props}>
      <FormLabel data-testid={LABEL_TESTID}>
        {LABEL_TEXT}
        <FormRequiredText data-testid={REQUIRED_TESTID}>
          {REQUIRED_TEXT}
        </FormRequiredText>
      </FormLabel>
      <Input data-testid={INPUT_TESTID} />
      <FormHelperText data-testid={HINT_TESTID}>{HINT_TEXT}</FormHelperText>
      <FormErrorText data-testid={ERROR_TESTID}>{ERROR_TEXT}</FormErrorText>
    </FormField>
  );
};

describe("<FormField />", () => {
  expect.assertions(1);

  it("should render FormField", () => {
    const { asFragment } = render(<FormFieldComponent id="email" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render FormErrorText when isInvalid is true", () => {
    const { rerender } = render(<FormFieldComponent isInvalid id="email" />);
    expect(screen.queryByTestId(ERROR_TESTID)).toBeInTheDocument();

    rerender(<FormFieldComponent id="email" />);
    expect(screen.queryByTestId(ERROR_TESTID)).not.toBeInTheDocument();
  });

  it("should render RequiredText when isRequired is true", () => {
    const { rerender } = render(<FormFieldComponent isRequired id="email" />);

    expect(screen.queryByTestId(REQUIRED_TESTID)).toBeInTheDocument();

    rerender(<FormFieldComponent id="email" />);
    expect(screen.queryByTestId(REQUIRED_TESTID)).not.toBeInTheDocument();
  });

  it("should set proper aria attributes to input on isDisabled", () => {
    const { rerender } = render(<FormFieldComponent isDisabled id="email" />);

    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute("disabled");
    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute(
      "aria-disabled",
      "true",
    );

    rerender(<FormFieldComponent id="email" />);
    expect(screen.queryByTestId(INPUT_TESTID)).not.toHaveAttribute("disabled");
  });

  it("should set proper aria attributes to input on isRequired", () => {
    const { rerender } = render(<FormFieldComponent isRequired id="email" />);

    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute("required");
    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute(
      "aria-required",
      "true",
    );

    rerender(<FormFieldComponent id="email" />);
    expect(screen.queryByTestId(INPUT_TESTID)).not.toHaveAttribute("required");
  });

  it("should set proper aria attributes to input on isInvalid", () => {
    const { rerender } = render(<FormFieldComponent isInvalid id="email" />);

    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute(
      "aria-invalid",
      "true",
    );

    rerender(<FormFieldComponent id="email" />);
    expect(screen.queryByTestId(INPUT_TESTID)).not.toHaveAttribute(
      "aria-invalid",
    );
  });

  it("should set proper aria attributes to input on isReadOnly", () => {
    const { rerender } = render(<FormFieldComponent isReadOnly id="email" />);

    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute("readonly");
    expect(screen.queryByTestId(INPUT_TESTID)).toHaveAttribute(
      "aria-readonly",
      "true",
    );

    rerender(<FormFieldComponent id="email" />);
    expect(screen.queryByTestId(INPUT_TESTID)).not.toHaveAttribute(
      "aria-readonly",
    );
  });

  it("should support render props", () => {
    const { asFragment } = render(
      <FormField id="email" isReadOnly isDisabled isInvalid isRequired>
        {({ inputProps, setHasErrorText, setHasHelpText, ...props }) => (
          <>
            <FormLabel>{LABEL_TEXT}</FormLabel>
            <input {...inputProps} />
            <FormHelperText>{HINT_TEXT}</FormHelperText>
            {JSON.stringify(props, null, 2)}
          </>
        )}
      </FormField>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not have a11y violations", async () => {
    await testA11y(<FormFieldComponent id="email" />);
  });
});
