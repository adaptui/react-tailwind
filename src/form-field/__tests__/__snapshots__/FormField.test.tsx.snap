// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> should render FormField 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col space-y-2"
    id="field-email"
    role="group"
  >
    <label
      class="text-xs text-gray-900 inline-flex items-center justify-between"
      data-testid="testid-label"
      for="email"
      id="field-email-label"
    >
      Form label
    </label>
    <div
      class="lib:relative lib:inline-flex"
    >
      <input
        aria-describedby="field-email-error-text field-email-help-text"
        class="lib:appearance-none lib:w-full lib:outline-none lib:px-2 lib:py-1 lib:text-xs lib:shadow-sm lib:font-medium lib:placeholder-text-gray-500 lib:border lib:h-6 lib:border-gray-300 hover:lib:border-gray-400 lib:rounded-md lib:text-gray-800 focus-visible:lib:border-blue-500 focus-visible:lib:shadow-input disabled:lib:bg-gray-100 disabled:lib:shadow-none disabled:lib:cursor-not-allowed disabled:lib:border-gray-300"
        data-testid="testid-input"
        id="email"
      />
    </div>
    <div
      class="text-xs text-gray-500 inline-flex items-center space-x-1 leading-4"
      data-testid="testid-hint"
      id="field-email-help-text"
    >
      This is hint text
    </div>
  </div>
</DocumentFragment>
`;

exports[`<FormField /> should support render props 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col space-y-2"
    id="field-email"
    role="group"
  >
    <label
      class="text-xs text-gray-900 inline-flex items-center justify-between"
      for="email"
      id="field-email-label"
    >
      Form label
    </label>
    <input
      aria-describedby="field-email-help-text"
      aria-invalid="true"
      aria-readonly="true"
      aria-required="true"
      disabled=""
      id="email"
      readonly=""
      required=""
    />
    <div
      class="text-xs text-gray-500 inline-flex items-center space-x-1 leading-4"
      id="field-email-help-text"
    >
      This is hint text
    </div>
    {
  "isRequired": true,
  "isInvalid": true,
  "isReadOnly": true,
  "isDisabled": true,
  "inputId": "email",
  "fieldId": "field-email",
  "labelId": "field-email-label",
  "errorTextId": "field-email-error-text",
  "helpTextId": "field-email-help-text",
  "hasHelpText": true,
  "hasErrorText": false
}
  </div>
</DocumentFragment>
`;
